File: src/webui/api/status.rs
// Copyright (C) 2026 Michael Wilson <mike@mdwn.dev>
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, version 3.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// this program. If not, see <https://www.gnu.org/licenses/>.
//

use axum::{extract::State, response::IntoResponse, Json};
use serde_json::json;

use super::super::server::WebUiState;
use crate::build_info;

/// GET /api/status — returns build info and hardware status.
pub(super) async fn get_status(State(state): State<WebUiState>) -> impl IntoResponse {
    let hardware = state.player.hardware_status();
    Json(json!({
        "build": {
            "version": build_info::VERSION,
            "git_hash": build_info::GIT_HASH,
            "build_time": build_info::BUILD_TIME,
        },
        "hardware": hardware,
    }))
}

#[cfg(test)]
mod test {
    use super::super::super::api;
    use super::super::test_helpers::*;
    use axum::body::Body;
    use http::StatusCode;
    use tower::ServiceExt;

    #[tokio::test]
    async fn get_status_returns_build_and_hardware() {
        let (state, _dir) = test_state();
        let app = api::router().with_state(state);

        let response = app
            .oneshot(
                http::Request::builder()
                    .method("GET")
                    .uri("/status")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);
        let body = response_body(response).await;
        let parsed: serde_json::Value = serde_json::from_str(&body).unwrap();

        // Build info is present.
        assert!(parsed["build"]["version"].is_string());
        assert!(parsed["build"]["git_hash"].is_string());
        assert!(parsed["build"]["build_time"].is_string());

        // Hardware section is present with all subsystems.
        assert!(parsed["hardware"]["init_done"].is_boolean());
        // Profile fields are present (may be null in test state).
        assert!(!parsed["hardware"]["hostname"].is_object());
        assert!(!parsed["hardware"]["profile"].is_object());
        assert!(parsed["hardware"]["audio"]["status"].is_string());
        assert!(parsed["hardware"]["midi"]["status"].is_string());
        assert!(parsed["hardware"]["dmx"]["status"].is_string());
        assert!(parsed["hardware"]["trigger"]["status"].is_string());
    }

    #[tokio::test]
    async fn get_status_no_devices_shows_not_connected() {
        let (state, _dir) = test_state();
        let app = api::router().with_state(state);

        let response = app
            .oneshot(
                http::Request::builder()
                    .method("GET")
                    .uri("/status")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        let body = response_body(response).await;
        let parsed: serde_json::Value = serde_json::from_str(&body).unwrap();

        // Test state has init_done=true and no devices, so all should be not_connected.
        assert_eq!(parsed["hardware"]["init_done"], true);
        assert_eq!(parsed["hardware"]["audio"]["status"], "not_connected");
        assert_eq!(parsed["hardware"]["midi"]["status"], "not_connected");
        assert_eq!(parsed["hardware"]["dmx"]["status"], "not_connected");
        assert_eq!(parsed["hardware"]["trigger"]["status"], "not_connected");
    }
}

File: src/webui/svelte/src/lib/ws/status.ts
// Copyright (C) 2026 Michael Wilson <mike@mdwn.dev>
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation, version 3.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// this program. If not, see <https://www.gnu.org/licenses/>.
//

import { derived, writable } from "svelte/store";
import { wsConnected } from "./stores";

export interface SubsystemStatus {
  status: "connected" | "initializing" | "not_connected" | "not_configured";
  name: string | null;
}

export interface ControllerStatus {
  kind: string;
  status: string;
  detail: string | null;
  error: string | null;
}

export interface StatusData {
  build: {
    version: string;
    git_hash: string;
    build_time: string;
  };
  hardware: {
    init_done: boolean;
    hostname: string | null;
    profile: string | null;
    audio: SubsystemStatus;
    midi: SubsystemStatus;
    dmx: SubsystemStatus;
    trigger: SubsystemStatus;
  };
  controllers: ControllerStatus[];
}

/** Worst-case across all required subsystems. */
export type Health = "ok" | "warn" | "error" | "unknown";

export const statusStore = writable<StatusData | null>(null);

let pollHandle: ReturnType<typeof setInterval> | null = null;

async function fetchStatus() {
  try {
    const res = await fetch("/api/status");
    if (!res.ok) return;
    const data = (await res.json()) as StatusData;
    statusStore.set(data);
  } catch {
    // Fetch failures are tolerated; the WS connect indicator already
    // surfaces total disconnects via the disconnect banner.
  }
}

function startPolling() {
  if (pollHandle !== null) return;
  void fetchStatus();
  pollHandle = setInterval(fetchStatus, 5000);
}

function stopPolling() {
  if (pollHandle === null) return;
  clearInterval(pollHandle);
  pollHandle = null;
}

// Poll while the WebSocket is up. When it drops we keep the last known
// snapshot but stop hammering /api/status.
wsConnected.subscribe((connected) => {
  if (connected) startPolling();
  else stopPolling();
});

/**
 * Derives the worst-case health across required subsystems.
 *
 * - Audio is always required.
 * - MIDI / DMX are required if the active profile has them configured
 *   (i.e. they have a `name`). A song-aware refinement (only required if
 *   the active song uses them) is a follow-up — for now, "configured but
 *   not connected" surfaces as amber.
 * - Trigger is never required.
 */
export const healthStore = derived(statusStore, ($status): Health => {
  if (!$status) return "unknown";

  const subs = $status.hardware;
  if (!subs.init_done) return "warn";

  // Audio is always required — its status is the floor.
  if (subs.audio.status === "not_connected") return "error";

  // MIDI / DMX are required if configured.
  for (const sub of [subs.midi, subs.dmx] as const) {
    if (sub.name && sub.status === "not_connected") return "error";
  }

  // Initializing or any controller in error → amber.
  for (const sub of [subs.audio, subs.midi, subs.dmx] as const) {
    if (sub.status === "initializing") return "warn";
  }
  if ($status.controllers.some((c) => c.status !== "running")) return "warn";

  return "ok";
});
